cwms-data-api/src/main/java/cwms/cda/data/dao/MeasurementDao.java

```diff
@@ -0,0 +1,500 @@
+import cwms.cda.data.dto.measurement.Measurement;
+import cwms.cda.data.dto.measurement.StreamflowMeasurement;
+import cwms.cda.data.dto.measurement.SupplementalStreamflowMeasurement;
+import cwms.cda.data.dto.measurement.UsgsMeasurement;
+import java.io.IOException;
+import java.sql.Timestamp;
+import java.time.Instant;
+import java.time.OffsetDateTime;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Date;
+import java.util.TimeZone;
+import mil.army.usace.hec.metadata.location.LocationTemplate;
+import org.jooq.DSLContext;
+
+import java.util.List;
+
+import static java.util.stream.Collectors.toList;
+import org.jooq.impl.DSL;
+import usace.cwms.db.dao.util.OracleTypeMap;
+import usace.cwms.db.jooq.codegen.packages.CWMS_STREAM_PACKAGE;
+import usace.cwms.db.jooq.codegen.udt.records.STREAMFLOW_MEAS2_T;
+import usace.cwms.db.jooq.codegen.udt.records.STREAMFLOW_MEAS2_TAB_T;
+
+public final class MeasurementDao extends JooqDao<Measurement> {
+    static final XmlMapper XML_MAPPER = buildXmlMapper();
+
+    public MeasurementDao(DSLContext dsl) {
+        super(dsl);
+    }
+
+    /**
+     * Retrieve a list of measurements
+     *
+     * @param officeId   - the office id
+     * @param locationId - the location id for filtering
+     * @param unitSystem - the unit system to use for the returned data
+     * @return a list of measurements
+     */
+    public List<Measurement> retrieveMeasurements(String officeId, String locationId, Instant minDateMask, Instant maxDateMask, String unitSystem) {
+        return connectionResult(dsl, conn -> {
+            setOffice(conn, officeId);
+            Number minHeight = null;
+            Number maxHeight = null;
+            Number minFlow = null;
+            Number maxFlow = null;
+            String minNum = null;
+            String maxNum = null;
+            String agencies = null;
+            String qualities = null;
+            Timestamp minTimestamp = OracleTypeMap.buildTimestamp(minDateMask == null ? null : Date.from(minDateMask));
+            Timestamp maxTimestamp = OracleTypeMap.buildTimestamp(maxDateMask == null ? null : Date.from(maxDateMask));
+            TimeZone timeZone = OracleTypeMap.GMT_TIME_ZONE;
+            STREAMFLOW_MEAS2_TAB_T retrieved = CWMS_STREAM_PACKAGE.call_RETRIEVE_MEAS_OBJS(DSL.using(conn).configuration(), locationId, unitSystem, minTimestamp, maxTimestamp,
+                    minHeight, maxHeight, minFlow, maxFlow, minNum, maxNum, agencies, qualities, timeZone.getID(), officeId);
+            return retrieved.stream()
+                    .map(MeasurementDao::fromJooqMeasurementRecord)
+                    .collect(toList());
+        });
+    }
+
+    /**
+     * Store a measurement
+     *
+     * @param measurement  - the measurement to store
+     * @param failIfExists - if true, fail if the measurement already exists
+     */
+    public void storeMeasurement(Measurement measurement, boolean failIfExists) {
```

Comment:
Why do we only have a store single measurement option instead of a list of measurements?  This is going to be incredibly slow if we're importing a thousand measurements.

Reply:
I don't think CDA supports bulk store calls anywhere else, and the db expects a single-store for measurements. From the hec-cwms-data-access perspective, I would iterate over a collection of measurements and make individual POST calls. 

Similar to how the measurement jdbc dao currently iterates over the set and make individual store calls into the jooq layer.

Reply:
As we discussed on Teams, please try doing a multi-store first, and if it works, please use that instead.  The DB does indicate it supports multiple storage.
Additionally, we should make sure we update the JDBC dao during HCDA updates to use the multi-store, because that would be an additional speed improvement and keep the behavior consistent.

Reply:
Tested out the bulk store and looks like it works successfully. Added unit tests and integrations tests.